presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
+                            child.getNode(),
+                            node.getPartitionBy(),
+                            node.getOrderBy(),
+                            node.getOrderings(),
+                            node.getRowNumberSymbol(),
+                            node.getMaxRowCountPerPartition(),
+                            node.isPartial(),
+                            Optional.of(hashSymbol)),
+                    child.getHashSymbols());
         }
 
         @Override
-        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        public PlanWithProperties visitJoin(JoinNode node, Context context)
         {
             List<JoinNode.EquiJoinClause> clauses = node.getCriteria();
+            if (clauses.isEmpty()) {
+                PlanWithProperties left = planAndEnforce(node.getLeft(), new Context());
+                // drop undesired hash symbols from build to save memory
+                PlanWithProperties right = planAndEnforce(node.getRight(), new Context(), true);
+
+                Map<List<Symbol>, Symbol> allHashSymbols = new HashMap<>();
+                allHashSymbols.putAll(right.getHashSymbols());
+                allHashSymbols.putAll(left.getHashSymbols());
+
+                return new PlanWithProperties(
+                        new JoinNode(
+                                idAllocator.getNextId(),
+                                node.getType(),
+                                left.getNode(),
+                                right.getNode(),
+                                node.getCriteria(),
+                                Optional.empty(),
+                                Optional.empty()),
+                        allHashSymbols);
+            }
 
+            // join does not pass through preferred hash symbols since they take more memory and since
+            // the join node filters, may take more compute
             List<Symbol> leftSymbols = Lists.transform(clauses, JoinNode.EquiJoinClause::getLeft);
+            PlanWithProperties left = planAndEnforce(node.getLeft(), new Context(leftSymbols));
+            Symbol leftHashSymbol = left.getRequiredHashSymbol(leftSymbols);
+
             List<Symbol> rightSymbols = Lists.transform(clauses, JoinNode.EquiJoinClause::getRight);
+            // drop undesired hash symbols from build to save memory
+            PlanWithProperties right = planAndEnforce(node.getRight(), new Context(rightSymbols), true);
+            Symbol rightHashSymbol = right.getRequiredHashSymbol(rightSymbols);
+
+            // build map of all hash symbols
+            Map<List<Symbol>, Symbol> allHashSymbols = new HashMap<>();
+            if (node.getType() == INNER || node.getType() == LEFT) {
+                allHashSymbols.putAll(left.getHashSymbols());
+            }
+            if (node.getType() == INNER || node.getType() == RIGHT) {
+                allHashSymbols.putAll(right.getHashSymbols());
+            }
```

Comment:
leave a comment here about how FULL type is handled, and why we don't need to mention it